db/batch/batch_impl.go

```diff
@@ -96,20 +96,42 @@ func (b *baseKVStoreBatch) Entry(index int) (*WriteInfo, error) {
 func (b *baseKVStoreBatch) SerializeQueue(serialize WriteInfoSerialize, filter WriteInfoFilter) []byte {
 	b.mutex.Lock()
 	defer b.mutex.Unlock()
-	// 1. This could be improved by being processed in parallel
-	// 2. Digest could be replaced by merkle root if we need proof
-	bytes := make([]byte, 0)
-	for _, wi := range b.writeQueue {
-		if filter != nil && filter(wi) {
-			continue
-		}
-		if serialize != nil {
-			bytes = append(bytes, serialize(wi)...)
-		} else {
-			bytes = append(bytes, wi.Serialize()...)
-		}
+	// 1. Digest could be replaced by merkle root if we need proof
+	var (
+		mutex           sync.Mutex
+		serialisedBytes = make([][]byte, len(b.writeQueue))
+		wg              = sync.WaitGroup{}
+	)
+
+	wg.Add(len(b.writeQueue))
+	for i, wi := range b.writeQueue {
+		go func(i int, info *WriteInfo) {
+			defer wg.Done()
+			if filter != nil && filter(info) {
+				return
+			}
+
+			idx := i
+			var data []byte
+			if serialize != nil {
+				data = serialize(info)
+			} else {
+				data = info.Serialize()
+			}
+
+			mutex.Lock()
+			serialisedBytes[idx] = data
+			mutex.Unlock()
```

Comment:
there is no need to lock here, as the concurrent writes on the `serializedBytes` occur at different indices.

Reply:
Thanks, didn't realise that. Fixed.